File: NowPlaying/iPhone/Classes/IMDbCache.m
// Copyright 2008 Cyrus Najmabadi
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "IMDbCache.h"

#import "Application.h"
#import "FileUtilities.h"
#import "LinkedSet.h"
#import "Movie.h"
#import "NetworkUtilities.h"
#import "NowPlayingAppDelegate.h"
#import "NowPlayingModel.h"
#import "ThreadingUtilities.h"
#import "Utilities.h"

@interface IMDbCache()
@property (retain) LinkedSet* normalMovies;
@property (retain) LinkedSet* prioritizedMovies;
@end


@implementation IMDbCache

@synthesize normalMovies;
@synthesize prioritizedMovies;

- (void) dealloc {
    self.normalMovies = nil;
    self.prioritizedMovies = nil;

    [super dealloc];
}


- (id) initWithModel:(NowPlayingModel*) model_ {
    if (self = [super initWithModel:model_]) {
        self.normalMovies = [LinkedSet set];
        self.prioritizedMovies = [LinkedSet setWithCountLimit:8];
        
        [ThreadingUtilities performSelector:@selector(updateBackgroundEntryPoint)
                                   onTarget:self
                       inBackgroundWithGate:nil
                                    visible:NO];
    }

    return self;
}


- (NSString*) imdbFile:(Movie*) movie {
    NSString* name = [[FileUtilities sanitizeFileName:movie.canonicalTitle] stringByAppendingPathExtension:@"plist"];
    return [[Application imdbDirectory] stringByAppendingPathComponent:name];
}


- (void) updateIMDbAddress:(Movie*) movie {
    if (movie.imdbAddress.length > 0) {
        // don't even bother if the movie has an imdb address in it
        return;
    }
    
    NSString* path = [self imdbFile:movie];
    NSDate* lastLookupDate = [FileUtilities modificationDate:path];
    
    if (lastLookupDate != nil) {
        NSString* value = [FileUtilities readObject:path];
        if (value.length > 0) {
            // we have a real imdb value for this movie
            return;
        }
        
        // we have a sentinel.  only update if it's been long enough
        if (ABS(lastLookupDate.timeIntervalSinceNow) < (3 * ONE_DAY)) {
            return;
        }
    }
    
    NSString* url = [NSString stringWithFormat:@"http://%@.appspot.com/LookupIMDbListings?q=%@", [Application host], [Utilities stringByAddingPercentEscapes:movie.canonicalTitle]];
    NSString* imdbAddress = [NetworkUtilities stringWithContentsOfAddress:url important:NO];
    if (imdbAddress == nil) {
        return;
    }
    
    // write down the response (even if it is empty).  An empty value will
    // ensure that we don't update this entry too often.
    [FileUtilities writeObject:imdbAddress toFile:path];
    if (imdbAddress.length > 0) {
        [NowPlayingAppDelegate minorRefresh];
    }
}


- (void) updateOneMovie {
    Movie* movie = nil;
    [gate lock];
    {
        while ((movie = [prioritizedMovies removeLastObjectAdded]) == NULL &&
               (movie = [normalMovies removeLastObjectAdded]) == NULL) {
            [gate wait];
        }
    }
    [gate unlock];
    
    [self updateIMDbAddress:movie];
}


- (void) updateBackgroundEntryPoint {
    while (YES) {
        NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
        {
            [self updateOneMovie];
        }
        [pool release];
    }
}


+ (IMDbCache*) cacheWithModel:(NowPlayingModel*) model {
    return [[[IMDbCache alloc] initWithModel:model] autorelease];
}


- (void) update:(NSArray*) movies {
    [gate lock];
    {
        [normalMovies addObjectsFromArray:movies];
        [gate signal];
    }
    [gate unlock];
}


- (NSSet*) cachedDirectoriesToClear {
    return [NSSet setWithObject:[Application imdbDirectory]];
}


- (void) prioritizeMovie:(Movie*) movie {
    [gate lock];
    {
        [prioritizedMovies addObject:movie];
        [gate signal];
    }
    [gate unlock];
}


- (NSString*) imdbAddressForMovie:(Movie*) movie {
    return [FileUtilities readObject:[self imdbFile:movie]];
}

@end

File: trunk/NowPlaying/Android/src/org/metasyntactic/caches/IMDbCache.java
//Copyright 2008 Cyrus Najmabadi
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.

package org.metasyntactic.caches;

import org.metasyntactic.Application;
import org.metasyntactic.Constants;
import org.metasyntactic.data.Movie;
import org.metasyntactic.threading.ThreadingUtilities;
import org.metasyntactic.utilities.FileUtilities;
import org.metasyntactic.utilities.NetworkUtilities;
import org.metasyntactic.utilities.StringUtilities;
import static org.metasyntactic.utilities.StringUtilities.isNullOrEmpty;

import java.io.File;
import java.util.*;

public class IMDbCache {
  private final Object lock = new Object();

  private String movieFileName(final Movie movie) {
    return FileUtilities.sanitizeFileName(movie.getCanonicalTitle());
  }

  private File movieFilePath(final Movie movie) {
    return new File(Application.imdbDirectory, movieFileName(movie));
  }

  public void update(final List<Movie> movies) {
    final Runnable runnable = new Runnable() {
      public void run() {
        updateBackgroundEntryPoint(movies);
      }
    };
    ThreadingUtilities.performOnBackgroundThread("Upate IMDb", runnable, this.lock, false);
  }

  private void updateBackgroundEntryPoint(final List<Movie> movies) {
    deleteObsoleteAddresses(movies);
    downloadImdbAddresses(movies);
  }

  private void deleteObsoleteAddresses(final List<Movie> movies) {
    final File imdbDir = Application.imdbDirectory;
    final Set<String> fileNames = new HashSet<String>(Arrays.asList(imdbDir.list()));

    for (final Movie movie : movies) {
      fileNames.remove(movieFileName(movie));
    }

    final long now = new Date().getTime();

    for (final String fileName : fileNames) {
      final File file = new File(imdbDir, fileName);
      if (file.exists()) {
        final long writeTime = file.lastModified();
        final long span = Math.abs(writeTime - now);

        if (span > 4 * Constants.ONE_WEEK) {
          file.delete();
        }
      }
    }
  }

  private void downloadImdbAddresses(final List<Movie> movies) {
    for (final Movie movie : movies) {
      final File path = movieFilePath(movie);
      if (path.exists()) {
        continue;
      }

      final String url = "http://" + Application
          .host + ".appspot.com/LookupIMDbListings?q=" + StringUtilities.urlEncode(movie.getCanonicalTitle());

      final String imdbAddress = NetworkUtilities.downloadString(url, false);

      if (!isNullOrEmpty(imdbAddress)) {
        FileUtilities.writeString(imdbAddress, movieFilePath(movie));
        Application.refresh();
      }
    }
  }

  public String imdbAddressForMovie(final Movie movie) {
    return FileUtilities.readString(movieFilePath(movie));
  }
}